File: test-btle-scan.py
#!/usr/bin/env python3
"""Test script for btle_scan with Rust USB backend."""

import json
import sys
import os

# Add the project root to the path
sys.path.insert(0, os.path.join(os.path.dirname(__file__), "python"))

from ubertooth_connector import UbertoothConnector

def main():
    print("=" * 60)
    print("Testing btle_scan with Rust USB backend")
    print("=" * 60)
    print()

    # Create connector instance
    connector = UbertoothConnector()

    # Test 1: Device connection
    print("[1/3] Connecting to Ubertooth device...")
    result = connector.execute("device_connect", {})

    if not result.get("success"):
        print(f"❌ Connection failed: {result.get('error', 'Unknown error')}")
        return 1

    print(f"✅ Connected to {result.get('board_name', 'Unknown')}")
    print(f"   Firmware: {result.get('firmware_version', 'Unknown')}")
    print(f"   Serial: {result.get('serial', 'Unknown')}")
    print()

    # Test 2: BLE scan (5 seconds)
    print("[2/3] Starting BLE scan (5 seconds on channel 37)...")
    print("      Looking for BLE advertisements...")

    scan_result = connector.execute("btle_scan", {
        "duration_sec": 5,
        "channel": 37,
        "save_pcap": True
    })

    if not scan_result.get("success"):
        print(f"❌ Scan failed: {scan_result.get('error', 'Unknown error')}")
        return 1

    print(f"✅ Scan completed!")
    print(f"   Duration: {scan_result.get('scan_duration_sec', 0)} seconds")
    print(f"   Channel: {scan_result.get('channel', 0)}")
    print(f"   Total packets: {scan_result.get('total_packets', 0)}")
    print(f"   Devices found: {len(scan_result.get('devices_found', []))}")
    print(f"   Capture ID: {scan_result.get('capture_id', 'N/A')}")

    if scan_result.get("pcap_path"):
        print(f"   PCAP path: {scan_result['pcap_path']}")

    print()

    # Show discovered devices
    devices = scan_result.get("devices_found", [])
    if devices:
        print("   Discovered BLE devices:")
        for i, device in enumerate(devices, 1):
            mac = device.get("mac_address", "Unknown")
            name = device.get("device_name", "Unknown")
            rssi = device.get("rssi_avg", 0)
            packets = device.get("packet_count", 0)
            print(f"     {i}. {mac}")
            print(f"        Name: {name}")
            print(f"        RSSI: {rssi} dBm")
            print(f"        Packets: {packets}")
    else:
        print("   ⚠️  No BLE devices detected (normal if no devices nearby)")

    print()

    # Test 3: Device disconnect
    print("[3/3] Disconnecting...")
    disconnect_result = connector.execute("device_disconnect", {})

    if disconnect_result.get("success"):
        print("✅ Disconnected successfully")
    else:
        print("⚠️  Disconnect warning:", disconnect_result.get("message", ""))

    print()
    print("=" * 60)
    print("✅ All tests completed successfully!")
    print("=" * 60)

    return 0

if __name__ == "__main__":
    sys.exit(main())

File: examples/test_btle_scan.rs
//! Test BLE scanning with native Rust USB backend

use serde_json::json;
use std::sync::Arc;
use tokio::sync::Mutex;
use ubertooth_usb::{UbertoothCommands, UbertoothDevice};

#[tokio::main]
async fn main() -> Result<(), Box<dyn std::error::Error>> {
    // Initialize logging
    tracing_subscriber::fmt()
        .with_env_filter("info,ubertooth_usb=debug")
        .init();

    println!("==============================================");
    println!("BLE Scan Test - Rust USB Backend");
    println!("==============================================");
    println!();

    // Create and connect to device
    println!("[1/4] Creating USB device connection...");
    let mut device = UbertoothDevice::new()?;

    println!("[2/4] Connecting to Ubertooth One (index 0)...");
    device.connect(0)?;

    let info = device.device_info().expect("Device info should be available");
    println!("✅ Connected to: {} (firmware: {})",
             info.board_name(),
             info.firmware_version);
    println!("   Board ID: {}", info.board_id);
    println!("   Serial: {}", info.serial_number);
    println!();

    // Create command executor
    println!("[3/4] Creating command executor...");
    let device = Arc::new(Mutex::new(device));
    let commands = UbertoothCommands::new(device.clone());
    println!("✅ Command executor ready");
    println!();

    // Run BLE scan
    println!("[4/4] Starting BLE scan...");
    println!("   Duration: 5 seconds");
    println!("   Channel: 37 (BLE advertising)");
    println!("   Mode: Promiscuous");
    println!();
    println!("Scanning for BLE advertisements...");
    println!();

    let start = std::time::Instant::now();

    let result = commands.btle_scan(json!({
        "duration_sec": 5,
        "channel": 37,
        "promiscuous": true,
        "save_pcap": false
    })).await;

    let elapsed = start.elapsed();

    match result {
        Ok(scan_result) => {
            println!();
            println!("==============================================");
            println!("✅ Scan Completed in {:.2}s", elapsed.as_secs_f64());
            println!("==============================================");
            println!();

            // Extract key metrics
            let total_packets = scan_result["total_packets"].as_u64().unwrap_or(0);
            let devices = scan_result["devices_found"].as_array();
            let device_count = devices.map(|d| d.len()).unwrap_or(0);
            let channel = scan_result["channel"].as_u64().unwrap_or(0);

            println!("Summary:");
            println!("  • Channel: {}", channel);
            println!("  • Total packets: {}", total_packets);
            println!("  • Unique devices: {}", device_count);
            println!("  • Scan time: {:.2}s", elapsed.as_secs_f64());

            if total_packets > 0 {
                println!("  • Throughput: {:.0} packets/sec",
                         total_packets as f64 / elapsed.as_secs_f64());
            }

            println!();

            // Show discovered devices
            if let Some(devices) = devices {
                if !devices.is_empty() {
                    println!("Discovered BLE Devices:");
                    println!();
                    for (i, device) in devices.iter().enumerate() {
                        let mac = device["mac_address"].as_str().unwrap_or("Unknown");
                        let name = device["device_name"].as_str().unwrap_or("Unknown");
                        let rssi = device["rssi_avg"].as_i64().unwrap_or(0);
                        let packets = device["packet_count"].as_u64().unwrap_or(0);

                        println!("  {}. {} ({})", i + 1, mac, name);
                        println!("     RSSI: {} dBm | Packets: {}", rssi, packets);
                    }
                    println!();
                } else {
                    println!("ℹ️  No BLE devices detected");
                    println!("   (This is normal if no BLE devices are nearby)");
                    println!();
                }
            }

            // Show raw result for debugging
            if std::env::var("SHOW_RAW").is_ok() {
                println!("Raw scan result:");
                println!("{}", serde_json::to_string_pretty(&scan_result)?);
                println!();
            }

            println!("==============================================");
            println!("✅ Test Successful!");
            println!("==============================================");
        }
        Err(e) => {
            println!();
            println!("==============================================");
            println!("❌ Scan Failed");
            println!("==============================================");
            println!();
            println!("Error: {}", e);
            println!();
            return Err(e.into());
        }
    }

    // Disconnect
    println!();
    println!("Disconnecting...");
    let mut device = device.lock().await;
    device.disconnect()?;
    println!("✅ Disconnected");

    Ok(())
}
